facilitator/src/bin/facilitator.rs

```diff
@@ -900,27 +1013,101 @@ fn aggregate(sub_matches: &ArgMatches) -> Result<(), anyhow::Error> {
         &mut aggregation_transport,
     )?;
 
+    aggregator.generate_sum_part(&parsed_batches)
+}
+
+fn aggregate_subcommand(sub_matches: &ArgMatches) -> Result<(), anyhow::Error> {
+    let batch_ids: Vec<&str> = sub_matches
+        .values_of("batch-id")
+        .context("no batch-id")?
+        .collect();
+    let batch_dates: Vec<&str> = sub_matches
+        .values_of("batch-time")
+        .context("no batch-time")?
+        .collect();
+
+    if batch_ids.len() != batch_dates.len() {
+        return Err(anyhow!(
+            "must provide same number of batch-id and batch-date values"
+        ));
+    }
+    let batch_info: Vec<_> = batch_ids.into_iter().zip(batch_dates).collect();
+
+    aggregate(
+        &sub_matches.value_of("aggregation-id").unwrap(),
+        sub_matches.value_of("aggregation-start").unwrap(),
+        sub_matches.value_of("aggregation-end").unwrap(),
+        batch_info,
+        sub_matches,
+    )
+}
+
+fn aggregate_worker(sub_matches: &ArgMatches) -> Result<(), anyhow::Error> {
     let aggregation_started: Counter = register_counter!(
         "aggregation_jobs_begun",
         "Number of aggregation jobs that started (on the facilitator side)"
     )
-    .unwrap();
-    aggregation_started.inc();
-
-    let result = aggregator.generate_sum_part(&batch_info);
-
+    .context("failed to register counter for started aggregations")?;
     let aggregation_finished = register_counter_vec!(
         "aggregation_jobs_finished",
         "Number of aggregation jobs that finished (on the facilitator side)",
         &["status"]
     )
-    .unwrap();
-    match result {
-        Ok(_) => aggregation_finished.with_label_values(&["success"]).inc(),
-        Err(_) => aggregation_finished.with_label_values(&["error"]).inc(),
+    .context("failed to register counter for finished aggregations")?;
+
+    let mut queue = aggregation_task_queue_from_args(sub_matches)?;
+
+    loop {
+        if let Some(task_handle) = queue.dequeue()? {
+            info!("dequeued task: {}", task_handle);
+            aggregation_started.inc();
+
+            let batches: Vec<(&str, &str)> = task_handle
+                .task
+                .batches
+                .iter()
+                .map(|b| (b.id.as_str(), b.time.as_str()))
+                .collect();
+            let result = aggregate(
+                &task_handle.task.aggregation_id,
+                &task_handle.task.aggregation_start,
+                &task_handle.task.aggregation_end,
+                batches,
+                sub_matches,
+            );
+
+            match result {
+                Ok(_) => {
+                    aggregation_finished.with_label_values(&["success"]).inc();
+                    queue.acknowledge_task(task_handle)?;
+                }
+                Err(err) => {
+                    error!("error while processing task {}: {:?}", task_handle, err);
+                    aggregation_finished.with_label_values(&["error"]).inc();
```

Comment:
This might be a place to nack the pubsub message or send it to a deadletter queue.